src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -369,210 +369,128 @@ public ExponentialHistogramData GetExponentialHistogramData()
-                    break;
-                }
+            Debug.Assert(this.mpComponents?.ExemplarReservoir != null, "ExemplarReservoir was null");
 
-            case AggregationType.LongSumIncomingCumulative:
-                {
-                    Interlocked.Exchange(ref this.runningValue.AsLong, number);
-                    break;
-                }
+            this.mpComponents!.ExemplarReservoir!.Offer(
+                new ExemplarMeasurement<long>(number, tags, explicitBucketHistogramBucketIndex));
+        }
 
-            case AggregationType.LongGauge:
-                {
-                    Interlocked.Exchange(ref this.runningValue.AsLong, number);
-                    break;
-                }
+        this.CompleteUpdate();
+    }
 
-            case AggregationType.Histogram:
-                {
-                    this.UpdateHistogram((double)number);
-                    break;
-                }
+    internal void Update(double number, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+    {
+        this.Update(number, out var explicitBucketHistogramBucketIndex);
 
-            case AggregationType.HistogramWithMinMax:
-                {
-                    this.UpdateHistogramWithMinMax((double)number);
-                    break;
-                }
+        var exemplarSampler = this.aggregatorStore.ExemplarSampler;
+        if (exemplarSampler.EarlySampleDecision ?? exemplarSampler.ShouldSampleDouble(number, tags))
+        {
+            Debug.Assert(this.mpComponents?.ExemplarReservoir != null, "ExemplarReservoir was null");
 
-            case AggregationType.HistogramWithBuckets:
-                {
-                    this.UpdateHistogramWithBuckets((double)number);
-                    break;
-                }
+            this.mpComponents!.ExemplarReservoir!.Offer(
+                new ExemplarMeasurement<double>(number, tags, explicitBucketHistogramBucketIndex));
+        }
 
-            case AggregationType.HistogramWithMinMaxBuckets:
-                {
-                    this.UpdateHistogramWithBucketsAndMinMax((double)number);
-                    break;
-                }
+        this.CompleteUpdate();
+    }
 
-            case AggregationType.Base2ExponentialHistogram:
-                {
-                    this.UpdateBase2ExponentialHistogram((double)number);
-                    break;
-                }
+    internal void TakeSnapshot(bool outputDelta)
+    {
+        this.Snapshot(outputDelta);
 
-            case AggregationType.Base2ExponentialHistogramWithMinMax:
-                {
-                    this.UpdateBase2ExponentialHistogramWithMinMax((double)number);
-                    break;
-                }
+        var exemplarReservoir = this.mpComponents?.ExemplarReservoir;
+        if (exemplarReservoir != null)
+        {
+            this.mpComponents!.Exemplars = this.mpComponents.ExemplarReservoir!.Collect();
```

Comment:
Might be a weird question, would it be possible that user wants longer collection cycle for ExemplarReservoir than the MetricReader's export cycle (since the Exemplar is observing trends for events)?
i.e.: not every MetricReader.OnCollect will trigger ExemplarReservoir.OnCollect
This behavior doesn't seem to be defined in spec, but theoretically can save cycles for users.

Reply:
That would be a big big change. Collected exemplars are tied to a MetricPoint. The spec would have to have some way to do an exemplar-only collect. Definitely not going to tackle that here but if you want to propose it at the spec-level go for it!